third_party/move/move-vm/runtime/src/native_functions.rs

```diff
@@ -370,6 +380,106 @@ impl<'a, 'b> LoaderContext<'a, 'b> {
             )
         })
     }
+
+    /// Resolves a function by module id and function id, with expected function type,
+    /// and return an abstract function which can be used to construct a closure value. This
+    /// invokes the configured loader and handles gas metering.
+    ///
+    /// If the function exists and is public, its type will be matched against the expected
+    /// type. Any type arguments will be instantiated in course of matching. Eventually,
+    /// this function guarantees that the return value has indeed the expected type, including
+    /// constraints on type parameters.
+    pub fn resolve_function(
+        &mut self,
+        module_id: &ModuleId,
+        fun_id: &IdentStr,
+        expected_ty: &Type,
+    ) -> PartialVMResult<Result<Box<dyn AbstractFunction>, FunctionResolutionError>> {
+        use FunctionResolutionError::*;
+        dispatch_loader!(&self.module_storage, loader, {
+            match loader.load_function_definition(
+                &mut self.gas_meter,
+                self.traversal_context,
+                module_id,
+                fun_id,
+            ) {
+                Ok((module, function)) => self.verify_function(module, function, expected_ty),
+                Err(e) if e.major_status() == StatusCode::FUNCTION_RESOLUTION_FAILURE => {
+                    Ok(Err(FunctionNotFound))
+                },
+                Err(e) => Err(e.to_partial()),
+            }
+        })
+    }
+
+    fn verify_function(
+        &mut self,
+        module: Arc<Module>,
+        func: Arc<Function>,
+        expected_ty: &Type,
+    ) -> PartialVMResult<Result<Box<dyn AbstractFunction>, FunctionResolutionError>> {
+        use FunctionResolutionError::*;
+        if !func.is_public() {
+            return Ok(Err(FunctionNotAccessible));
+        }
+        let Type::Function {
+            args,
+            results,
+            // Since resolved functions must be public, they always have all possible
+            // abilities (store, copy, and drop), and we don't need to check with
+            // expected abilities.
+            abilities: _,
+        } = expected_ty
+        else {
+            return Ok(Err(FunctionIncompatibleType));
+        };
+        let func_ref = func.as_ref();
+
+        // Match types, inferring instantiation of function in `subst`.
+        let mut subst = TypeParamMap::default();
```

Comment:
This is the same as for constructor function in arg validation? Probably better to have a single utility?

Reply:
No, we still must check the argument and result types. The substitution is empty if the function is not generic. I added an expllicit check for the arity vs type parameter count.